python/pyarrow/dataset.py

```diff
@@ -52,3 +52,59 @@
     SchemaPartitionScheme,
     TreeDataSource,
 )
+
+
+def open_dataset(path_or_paths, filesystem=None, partition_scheme=None,
+                 format="parquet"):
+    """
+    Open a multi-file dataset.
+
+    Parameters
+    ----------
+    path_or_paths : str or list of str
+        String path to a directory containing the data files.
+    filesystem : FileSystem, default None
+        By default will be inferred from the path.
+    partition_scheme : PartitionScheme or Schema
+        If a Schema is passed, it is interpreted as a HivePartitionScheme.
```

Comment:
FWIW, this is a little different from what R does: https://github.com/apache/arrow/blob/master/r/R/dataset.R#L23-L29

Also, I think this can/should change when #6008 lands. IIUC discovery will automatically identify hive partitions. You may also want to accept a list of strings as the names of the (unnamed, non-hive) partition segments (or perhaps this is to be done on https://issues.apache.org/jira/browse/ARROW-7413).